rethinkdb/tests/cluster.py

```diff
@@ -0,0 +1,129 @@
+        assert response['ready'] == 1
+
+        # Users.
+        # See: https://rethinkdb.com/docs/permissions-and-accounts/
+        response = r.db('rethinkdb').table('users').insert({'id': AGENT_USER, 'password': AGENT_PASSWORD}).run(conn)
+        assert response['inserted'] == 1
+        response = r.db('rethinkdb').grant(AGENT_USER, {'read': True}).run(conn)
+        assert response['granted'] == 1
+
+        response = r.db('rethinkdb').table('users').insert({'id': CLIENT_USER, 'password': False}).run(conn)
+        assert response['inserted'] == 1
+        response = r.db(DATABASE).grant(CLIENT_USER, {'read': True, 'write': True}).run(conn)
+        assert response['granted'] == 1
+
+    # Simulate client activity.
+    # NOTE: ensures that 'written_docs_*' and 'read_docs_*' metrics have non-zero values.
+
+    with r.connect(host=HOST, port=SERVER_PORTS['proxy'], user=CLIENT_USER) as conn:
+        response = r.db(DATABASE).table(HEROES_TABLE).insert(HEROES_TABLE_DOCUMENTS).run(conn)
+        assert response['inserted'] == len(HEROES_TABLE_DOCUMENTS)
+
+        documents = list(r.db(DATABASE).table(HEROES_TABLE).run(conn))
+        assert len(documents) == len(HEROES_TABLE_DOCUMENTS)
+
+
+@contextmanager
+def temporarily_disconnect_server(server):
+    # type: (str) -> Iterator[None]
+    """
+    Gracefully disconnect a server from the cluster.
+
+    Ensures that the cluster/replicas are in a stable state (not rebalancing) inside and after exiting the context.
+    """
+    service = 'rethinkdb-{}'.format(server)
+    logger.debug('temporarily_disconnect_server server=%r service=%r', server, service)
+
+    def _server_exists(conn):
+        # type: (rethinkdb.net.Connection) -> bool
+        servers = r.db('rethinkdb').table('server_status').map(r.row['name']).run(conn)  # type: List[str]
+        logger.debug('server_exists server=%r servers=%r', server, servers)
+        return server in servers
+
+    def _leader_election_done(conn):
+        # type: (rethinkdb.net.Connection) -> bool
+        STABLE_REPLICA_STATES = {'ready', 'waiting_for_primary', 'disconnected'}
+
+        replica_states = list(
+            r.db('rethinkdb')
+            .table('table_status')
+            .concat_map(r.row['shards'])
+            .concat_map(r.row['replicas'])
+            .map(r.row['state'])
+            .run(conn)
+        )  # type: List[str]
+
+        logger.debug('replica_states %r', replica_states)
+
+        return all(state in STABLE_REPLICA_STATES for state in replica_states)
+
+    def _server_disconnected(conn):
+        # type: (rethinkdb.net.Connection) -> bool
+        return not _server_exists(conn) and _leader_election_done(conn)
+
+    def _server_reconnected(conn):
+        # type: (rethinkdb.net.Connection) -> bool
+        return _server_exists(conn) and _leader_election_done(conn)
+
+    with temporarily_stop_service(service, compose_file=COMPOSE_FILE):
+        with EnvVars(COMPOSE_ENV_VARS):
```

Comment:
Shouldn't this be outside `temporarily_stop_service`?

Reply:
Theoretically yes, but actually this is useless since env vars are already injected by the `dd_environment` fixture… Dropping